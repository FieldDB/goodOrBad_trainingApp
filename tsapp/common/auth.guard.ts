import { Injectable } from '@angular/core';
import { Router, CanActivate } from '@angular/router';


@Injectable()
export class AuthGuard implements CanActivate {
    constructor(private router: Router) { }

    canActivate() {
        // Here change that to the logic you want, cookie stuff, permission, etc.
        if (JSON.parse(localStorage.getItem('goodOrBadUser')).role === 'manager') {
            return true;
        }

        this.router.navigate(['/login']);
        return false;
    }
}
