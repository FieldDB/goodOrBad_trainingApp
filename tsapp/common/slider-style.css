.default_null_value::-webkit-slider-runnable-track {
  background: #99ccff !important;
}

.slider_passing_value::-webkit-slider-runnable-track {
  background: green !important;
}

.slider_failing_value::-webkit-slider-runnable-track {
  background: red !important;
}

input[type=range].criteriaslider {
  -webkit-appearance: none;
  width: 100%;
  margin: 8px 0;
}
input[type=range].criteriaslider:focus {
  outline: none;
}
input[type=range].criteriaslider::-webkit-slider-runnable-track {
  width: 100%;
  height: 12px;
  cursor: pointer;
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
  background: #797979;
  border-radius: 2px;
  border: 0.2px solid #010101;
}
input[type=range].criteriaslider::-webkit-slider-thumb {
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
  border: 1px solid #000000;
  height: 28px;
  width: 14px;
  border-radius: 25px;
  background: rgba(255, 34, 1, 0.77);
  cursor: pointer;
  -webkit-appearance: none;
  margin-top: -8.2px;
}
input[type=range].criteriaslider:focus::-webkit-slider-runnable-track {
  background: #b6b6b6;
}
input[type=range].criteriaslider::-moz-range-track {
  width: 100%;
  height: 12px;
  cursor: pointer;
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
  background: #797979;
  border-radius: 2px;
  border: 0.2px solid #010101;
}
input[type=range].criteriaslider::-moz-range-thumb {
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
  border: 1px solid #000000;
  height: 28px;
  width: 14px;
  border-radius: 25px;
  background: rgba(255, 34, 1, 0.77);
  cursor: pointer;
}
input[type=range].criteriaslider::-ms-track {
  width: 100%;
  height: 12px;
  cursor: pointer;
  background: transparent;
  border-color: transparent;
  color: transparent;
}
input[type=range].criteriaslider::-ms-fill-lower {
  background: #3c3c3c;
  border: 0.2px solid #010101;
  border-radius: 4px;
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
}
input[type=range].criteriaslider::-ms-fill-upper {
  background: #797979;
  border: 0.2px solid #010101;
  border-radius: 4px;
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
}
input[type=range].criteriaslider::-ms-thumb {
  box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
  border: 1px solid #000000;
  height: 28px;
  width: 14px;
  border-radius: 25px;
  background: rgba(255, 34, 1, 0.77);
  cursor: pointer;
  height: 12px;
}
input[type=range].criteriaslider:focus::-ms-fill-lower {
  background: #797979;
}
input[type=range].criteriaslider:focus::-ms-fill-upper {
  background: #b6b6b6;
}
