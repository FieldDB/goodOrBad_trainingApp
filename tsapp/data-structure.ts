export class GoldenRow {
    filename: string;
    url: string;
    description: string;
    criteria_array: string;
    criteria_array_converted: number[];
    creation_date: string;
    passfail: boolean;
    explanation: string;
    type: string;
    info_url: string;
    oid: number;
}

export class ResultValue {
    username: string;
    filenameid: number;
    golden_passfail_state: boolean;
    success: boolean;
    fail_passed: boolean;
    positive_failed: boolean;
    delta_criteria_array: string;
    delta_array: number[];
    inspection_date: string;
    user_comments: string;
    type: string;
    timeinsec: number;
    oid: number;
}

export class CriteriaObject {
    crit: string;
    active: boolean; // Impossible to predict everything, so eventually in the future we want to remove some criteria.
    // Option1: Having the DB have 1 cell per criteria --> If I want to add or remove it I have to change the DB and the code. Bad.
    // Option2 (currently selected): Having the criteria as an array in the cell -->
    // Fast to add new one and infite amount, Bad for sorting, Bad if we want to remove some on the way.
    // Since we have option2 we add flexibility by adding a Show/hide flag
    // Option3 would be to take option2 and instead of array of value/null we could strignify a object to have key-value pair,
    // but it is heavier on the DB.;
}

export class PassFailResult {
    username: string;
    success: boolean;
    fail_passed: boolean;
    positive_failed: boolean;
    ct: number;
}

export class ResultFailRatio {
    username: string;
    success: boolean;
    golden_passfail_state: boolean;
    ct: number;
}

export class RatioPerUserStr {
    username: string;
    successgood: string; // this is a string since it come from the API as JSON but should eb a number
    successbad: string;
    failedgood: string;
    failedbad: string;
    total: string;
}

export class RatioPerUser {
    username: string;
    successgood: number;
    successbad: number;
    failedgood: number;
    failedbad: number;
    total: number;
}

export class RatioPerUserPc {
    // Percentage result for each user.
    username: string;
    globSuccessRate: number;
    succOnGoodImg: number;
    succOnBadImg: number;
    total: number;
}

export class UserDetails {
  email: string;
  username: string;
  name: string;
}
