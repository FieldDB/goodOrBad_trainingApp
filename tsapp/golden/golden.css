.ng-invalid {
  border-left: 5px solid #a94442; /* red */
}

.form-description-label {
	min-width: 150px;
	text-align: left;
}

textarea.form-control {
	min-height: 75px;
}

.validationbtn {
	padding: 10px;
	min-width: 250px;
}