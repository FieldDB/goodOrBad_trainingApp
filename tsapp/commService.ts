import { Injectable }     from '@angular/core';
import { Http, Response, Headers, RequestOptions } from '@angular/http';
import { GoldenRow, ResultValue,PassFailResult, ResultFailRatio } from './data-structure';
import { Observable }     from 'rxjs/Observable';
// import 'rxjs/Rx';
// import 'rxjs/add/operator/toPromise';
// import 'rxjs/add/operator/';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';

@Injectable()
export class CommService {
  constructor (private http: Http) {}
  private baseUrl:string = 'http://localhost:8010/';
// Result page
  private randgoldenUrl = this.baseUrl + 'api/view/randgolden/';  // URL to web api
  getDbImg() : Observable<GoldenRow[]> {
    return this.http.get(this.randgoldenUrl)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  private resultUrl = this.baseUrl + 'api/result';  // URL to web api
  postDbResult (result: ResultValue): Observable<ResultValue> {
    let body = JSON.stringify(result);
    let headers = new Headers({ 'Content-Type': 'application/json' });
    let options = new RequestOptions({ headers: headers });

    return this.http.post(this.resultUrl, body, options)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  // Golden Sample page
  getAllGolden() : Observable<GoldenRow[]> {
    let goldenImg:string = this.baseUrl + 'api/golden/';
    return this.http.get(goldenImg)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  getPreciceGolden(imgOid:number) : Observable<GoldenRow[]> {
    let goldenImg:string = this.baseUrl + 'api/golden/' + imgOid;
    return this.http.get(goldenImg)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  postNewGoldenImg (result: GoldenRow): Observable<GoldenRow> {
    let body = JSON.stringify(result);
    let headers = new Headers({ 'Content-Type': 'application/json' });
    let options = new RequestOptions({ headers: headers });

    return this.http.post(this.baseUrl + 'api/golden', body, options)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  updateGoldenImg (result: GoldenRow, oid: number): Observable<GoldenRow> {
    let body = JSON.stringify(result);
    let headers = new Headers({ 'Content-Type': 'application/json' });
    let options = new RequestOptions({ headers: headers });

    return this.http.post(this.baseUrl + 'api/golden/' + oid, body, options)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  // Differents API Views Querry
  // TODO: Refactor of ALL this.
  passfailresult(username?: string) : Observable<PassFailResult[]> {
    let getUrl:string = this.baseUrl + 'api/view/passfailresult/';
    if(username) {getUrl = getUrl + username;}
    return this.http.get(getUrl)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  resFailRatio(username?: string) : Observable<ResultFailRatio[]> {
    let getUrl:string = this.baseUrl + 'api/view/totimgfailratio/';
    if(username) {getUrl = getUrl + username;}
    return this.http.get(getUrl)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  private extractData(res: Response) {
    let body = res.json();
    return body || { };
  }

  private handleError (error: any) {
    // In a real world app, we might use a remote logging infrastructure
    // We'd also dig deeper into the error to get a better message
    let errMsg = (error.message) ? error.message :
      error.status ? `${error.status} - ${error.statusText}` : 'Server error';
    console.error(errMsg); // log to console instead
    return Observable.throw(errMsg);
  }
}
