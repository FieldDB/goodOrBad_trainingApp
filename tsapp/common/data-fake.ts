// Returning some fake data to who need it.
import { CriteriaObject, GoldenRow } from '../data-structure';

export const ArrOfCriteria: CriteriaObject[] = [
    { crit: 'Does it contain Fur', active: true },
    { crit: 'Cuteness Criteria', active: true },
    { crit: 'Realistic level', active: true },
    { crit: 'How blue is the Background', active: true },
    { crit: 'Old criteria that we dont use anymore, (not active)', active: false },
    { crit: 'How old is the Kitten Years', active: true },
    { crit: 'Is the Kitten Grey?', active: true },
    { crit: 'Would you adopt the kitten?', active: true },
    { crit: 'Is there a Logo in the image?', active: true },
    { crit: 'Do you think the Kitten is planning something Evil?', active: true }
];

export const BlankGoldenImg: GoldenRow = {
    filename: '',
    url: '',
    description: '',
    criteria_array: '',
    criteria_array_converted: [],
    creation_date: '',
    passfail: null,
    explanation: '',
    type: '',
    info_url: '',
    info_url_arr: [],
    oid: null
};
