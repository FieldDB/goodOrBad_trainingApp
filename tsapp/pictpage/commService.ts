import { Injectable }     from '@angular/core';
import { Http, Response } from '@angular/http';
import { GoldenRow }      from '../data-structure';
import { Observable }     from 'rxjs/Observable';
import 'rxjs/Rx';
// import 'rxjs/add/operator/toPromise';
// import 'rxjs/add/operator/';
// import 'rxjs/add/operator/map';
// import 'rxjs/add/operator/catch';

@Injectable()
export class CommService {
  constructor (private http: Http) {}
  private randgoldenUrl = 'http://localhost:8010/api/view/randgolden';  // URL to web api
  
  getDbImg() : Observable<GoldenRow[]> {
    return this.http.get(this.randgoldenUrl)
                    .map(this.extractData)
                    .catch(this.handleError);
  }

  private extractData(res: Response) {
    let body = res.json();
    return body || { };
  }

  private handleError (error: any) {
    // In a real world app, we might use a remote logging infrastructure
    // We'd also dig deeper into the error to get a better message
    let errMsg = (error.message) ? error.message :
      error.status ? `${error.status} - ${error.statusText}` : 'Server error';
    console.error(errMsg); // log to console instead
    return Observable.throw(errMsg);
  }
}