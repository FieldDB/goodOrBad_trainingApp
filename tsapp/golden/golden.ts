import { Component, OnInit} from '@angular/core';
import { NgIf, NgFor, NgClass} from '@angular/common';

import { DefaultDataService } from '../data.service';
import { GoldenRow, CriteriaObject } from '../data-structure';
import { CommService } from '../commService';

@Component({
	selector: 'golden',
	directives: [NgIf, NgFor, NgClass],
	styleUrls: ['tsapp/golden/golden.css', 'tsapp/common/slider-style.css'],
	templateUrl: 'tsapp/golden/golden.html',
	providers: [ CommService, DefaultDataService ]
})

export class Golden implements OnInit{
	goldenDetails: GoldenRow;
	criterialist: CriteriaObject[];

	constructor(private commService: CommService, 
		private defaultDataService: DefaultDataService) {}
	
	ngOnInit() {
		this.getDefaultCriteria();
		this.resetBlankImg();
	}

	resetBlankImg() {
	  	this.defaultDataService.blankGoldenImg().then(data => {
	  		this.goldenDetails = data;
		});

	}

	getDefaultCriteria() {
	  this.defaultDataService.arrOfCriteria().then(data => this.criterialist = data);
	}

	fetchOnEnter(event, oid) {
	  if(event.keyCode == 13 && oid) {
	    this.getPreciceImg(oid);
	  }
	}

	private getPreciceImg(imgOid:number) {
		this.commService.getPreciceGolden(imgOid)
				.subscribe((arrayOf1:GoldenRow[]) => {
					if(arrayOf1[0]) {
						this.goldenDetails = arrayOf1[0];
						this.goldenDetails.criteria_array_converted = JSON.parse(this.goldenDetails.criteria_array);
					} else {
						this.resetBlankImg();
					}
				},
				error => {
  	               	console.log("ERROR:", error);
  	               	// Warn the user and display the Oid red?
  	            })
	}

	onSubmit() {
		console.log("Submitting the form");
		this.goldenDetails.criteria_array = JSON.stringify(this.goldenDetails.criteria_array_converted);

	}

	setRandomKittenUrl() {
		let nbr1:number = Math.floor(Math.random()*200) + 250; //Nbr between 250 and 450px
		let nbr2:number = Math.floor(Math.random()*200) + 250; //Nbr between 250 and 450px
		let grey:string = Math.random() > 0.5 ? 'g/' : '';
		this.goldenDetails.url = "https://placekitten.com/" + grey + nbr1 + "/" + nbr2;
	}
}
