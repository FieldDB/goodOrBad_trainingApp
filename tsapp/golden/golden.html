<div>
Option1: Get a Golden image from the system and modify it. Or Delete it (Add Delete flag in Node/postgres for Golden) <br>

To Add, if the OID is in the route load it directly. <br>
Add a "send" on enter key.
<div *ngIf='goldenDetails && criterialist'>
<div class="row">
	<div class="form-group col-sm-6">
		<div class="input-group">
			<div disabled="!oid.value" class="input-group-addon">Img #</div>
		  <input class="form-control" type="number" placeholder="oid# 16445" min=0 #oid (keydown)="fetchOnEnter($event, oid.value)">
		</div>
	</div>
	<div class="col-sm-6 text-center">
		<button class="btn btn-default" [ngClass]="{'btn-info': goldenDetails.oid}" (click)="resetBlankImg()">Reset to Blank</button>
	</div>
</div>
<br>
<form (ngSubmit)="onSubmit()" #goldenImg="ngForm">
	<div class="row">
		<div class="col-sm-4">
			<div class="row">
			  <div class="form-group col-xs-12">
			    <div class="input-group">
			      <div class="input-group-addon">URL</div>
			      <input type="url" class="form-control" required
			      [(ngModel)]="goldenDetails.url">
			    </div>
			  </div>
			</div>
			<div class="row text-center">
				<button (click)="setRandomKittenUrl()">Surprise me...</button>
			</div>
			<div class="row" *ngIf="goldenDetails.url">
				<a href="{{goldenDetails.url}}" target="_blank">
					<img class="col-xs-12" src="{{goldenDetails.url}}">
				</a>
			</div>
		</div>
		<div class="col-sm-8">
			<div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Filename</div>
		      <input type="text" class="form-control" required
		      placeholder="Top title of the image"
		      [(ngModel)]="goldenDetails.filename">
		    </div>
		  </div>
		  

		  <div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Pass or Fail</div>
		      <input type="button" class="form-control" required
		      (click)="goldenDetails.passfail = !goldenDetails.passfail"
		      [(ngModel)]="goldenDetails.passfail"
		      [ngClass]="{'btn-success':goldenDetails.passfail == true, 'btn-danger': goldenDetails.passfail == false}">
		    </div>
		  </div>

		  <div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Img type</div>
		      <input type="text" class="form-control"
		      placeholder="trademark focus contrast typo..."
		      [(ngModel)]="goldenDetails.type">
		    </div>
		  </div>

		  <div *ngFor='let criteria of criterialist; let i = index;'>
		  	{{criteria.crit}} <strong *ngIf="!criteria.active" class="bg-danger">- Disabled</strong>
		  	<div class="row text-center" *ngIf="goldenDetails.criteria_array_converted && criteria.active">
		  		<div class="col-xs-1">
		  			<i *ngIf='goldenDetails.criteria_array_converted[i] != null && goldenDetails.criteria_array_converted[i] <=5'
		  				class="glyphicon glyphicon-fire text-danger"></i></div>
		  		<div class="col-xs-1">{{goldenDetails.criteria_array_converted[i]}}</div>
			  	<div class="col-xs-8">
			  	<input 
			  		class="criteriaslider"
			  		[ngClass]="{'default_null_value':!criteria.active || goldenDetails.criteria_array_converted[i] == null}"
			  		min=0,
			  		max=10
			  		steps=1
			  		type="range"
			  		[disabled]="!criteria.active"
			  		[(ngModel)]="goldenDetails.criteria_array_converted[i]">
			  	</div>
		  		<div class="col-xs-1"><i *ngIf='goldenDetails.criteria_array_converted[i] >5'class="glyphicon glyphicon-ok text-success"></i></div>
		  		<div class="col-xs-1" (click)='goldenDetails.criteria_array_converted[i] = null'><i *ngIf="goldenDetails.criteria_array_converted[i]" class="glyphicon glyphicon-remove"></i></div>
		  		
		  	</div>
		  </div>
		  <br>

		  <div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Extra Info Url</div>
		      <input type="url" class="form-control"
		      [(ngModel)]="goldenDetails.info_url">
		    </div>
		  </div>

		  <div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Description</div>
		      <textarea type="text" class="form-control"
		      placeholder="Description to help the user make a decision"
		      [(ngModel)]="goldenDetails.description"></textarea>
		    </div>
		  </div>

		  <div class="form-group">
		    <div class="input-group">
		      <div class="input-group-addon form-description-label">Explanation</div>
		      <textarea type="text" class="form-control"
		      placeholder="Why it pass or fail?"
		      [(ngModel)]="goldenDetails.explanation"></textarea>
		    </div>
		  </div>



		</div>
	</div>
</form>

<pre>{{goldenDetails | json}}</pre>
</div>
</div>

